cpp/src/stream_compaction/distinct_count.cu

```diff
@@ -131,42 +108,167 @@ struct has_nans {
+ */
+struct check_nan {
+  /**
+   * @brief Checks if the row `index` of `input` is `NAN`.
+   *
+   * @note This will be applicable only for floating point type columns.
+   *
+   * @param[in] input The `column_device_view` which will be checked for `NAN`
+   * @param[in] index The index at which the `NAN` needs to be checked in `input`
+   *
+   * @returns bool true if value at `index` is `NAN`, else false
+   */
+  template <typename T, std::enable_if_t<std::is_floating_point<T>::value>* = nullptr>
+  __device__ bool operator()(column_device_view const& input, size_type index)
+  {
+    return std::isnan(input.data<T>()[index]);
+  }
+  /**
+   * @brief Checks if the row `index` of `input` is `NAN`.
+   *
+   * @note This will be applicable for non-floating point type columns. And
+   * non-floating point columns can never have `NAN`, so it will always return
+   * false.
+   *
+   * @param[in] input The `column_device_view` which will be checked for `NAN`
+   * @param[in] index The index at which the `NAN` needs to be checked in `input`
+   *
+   * @returns bool true if value at `index` is `NAN`, else false
+   */
+  template <typename T, std::enable_if_t<not std::is_floating_point<T>::value>* = nullptr>
+  __device__ bool operator()(column_device_view const&, size_type)
   {
     return false;
   }
 };
+}  // namespace
+
+cudf::size_type distinct_count(table_view const& keys,
+                               null_equality nulls_equal,
+                               rmm::cuda_stream_view stream)
+{
+  auto table_ptr = cudf::table_device_view::create(keys, stream);
+  row_equality_comparator comp(
+    nullate::DYNAMIC{cudf::has_nulls(keys)}, *table_ptr, *table_ptr, nulls_equal);
+  return thrust::count_if(
+    rmm::exec_policy(stream),
+    thrust::counting_iterator<cudf::size_type>(0),
+    thrust::counting_iterator<cudf::size_type>(keys.num_rows()),
+    [comp] __device__(cudf::size_type i) { return (i == 0 || not comp(i, i - 1)); });
+}
+
+cudf::size_type unordered_distinct_count(table_view const& keys,
+                                         null_equality nulls_equal,
+                                         rmm::cuda_stream_view stream)
+{
+  auto table_ptr      = cudf::table_device_view::create(keys, stream);
+  auto const num_rows = table_ptr->num_rows();
+  auto const has_null = nullate::DYNAMIC{cudf::has_nulls(keys)};
+
+  hash_map_type key_map{compute_hash_table_size(num_rows),
+                        COMPACTION_EMPTY_KEY_SENTINEL,
+                        COMPACTION_EMPTY_VALUE_SENTINEL,
+                        detail::hash_table_allocator_type{default_allocator<char>{}, stream},
+                        stream.value()};
+
+  compaction_hash hash_key{has_null, *table_ptr};
+  row_equality_comparator row_equal(has_null, *table_ptr, *table_ptr, nulls_equal);
+  auto iter = cudf::detail::make_counting_transform_iterator(
+    0, [] __device__(size_type i) { return cuco::make_pair(std::move(i), std::move(i)); });
```

Comment:
This is an issue we need to fix in `cuco`, Embarassingly, `cuco::make_pair(i, i)` cannot work.

Reply:
Hmm, okay. In the meantime, you can just create a second variable as a workaround to avoid UB:
```suggestion
    0, [] __device__(size_type i) {
        // TODO: cuco::make_pair currently requires rvalue references. We
        // create a copy to avoid double-move invoking undefined behavior.
        auto const i2 = i;
        return cuco::make_pair(std::move(i), std::move(i2));
    });
```

Reply:
Update: This is fixed by https://github.com/NVIDIA/cuCollections/pull/138